### ISSUE TITLE
ibm/onesided/c_accumulate and friends hang with osc rdma using openib BTL

### ISSUE THREAD

**Issue by rolfv** [open-mpi/ompi#1077]
With the master using --mca osc rdma, I see a variety of onesided tests that are hanging. The list of tests are: c_accumulate, c_accumulate_atomic, c_create_dynamic, c_fence_lock, c_flush, c_get_accumulate, c_post_start, c_reqops, c_strided_acc_indexed, etc....

Am I the only one that sees this?

```
[rvandevaart@drossetti-ivy4 onesided]$ mpirun --mca osc pt2pt --mca btl self,sm,openib --mca btl_openib_warn_default_gid_prefix 0 -np 2 -host drossetti-ivy4,drossetti-ivy5 c_accumulate
[rvandevaart@drossetti-ivy4 onesided]$ mpirun --mca osc rdma --mca btl self,sm,openib --mca btl_openib_warn_default_gid_prefix 0 -np 2 -host drossetti-ivy4,drossetti-ivy5 c_accumulate
....hang....
```

```
[rvandevaart@drossetti-ivy4 src]$ pstack 29329
Thread 3 (Thread 0x7fefc03fb700 (LWP 29330)):
#0  0x00007fefc28b3173 in epoll_wait () from /lib64/libc.so.6
#1  0x00007fefc2236506 in epoll_dispatch () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#2  0x00007fefc223bcc6 in opal_libevent2022_event_base_loop () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#3  0x00007fefc041d44d in progress_engine () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libpmix.so.0
#4  0x00007fefc2b659d1 in start_thread () from /lib64/libpthread.so.0
#5  0x00007fefc28b2b7d in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7fefbf3fa700 (LWP 29331)):
#0  0x00007fefc28a9353 in poll () from /lib64/libc.so.6
#1  0x00007fefc2241d1a in poll_dispatch () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#2  0x00007fefc223bcc6 in opal_libevent2022_event_base_loop () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#3  0x00007fefc21e0b40 in progress_engine () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#4  0x00007fefc2b659d1 in start_thread () from /lib64/libpthread.so.0
#5  0x00007fefc28b2b7d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7fefc3308700 (LWP 29329)):
#0  0x00007fefbdbeacc2 in opal_list_remove_item () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_rcache_vma.so
#1  0x00007fefbdbec841 in mca_rcache_vma_tree_delete () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_rcache_vma.so
#2  0x00007fefbdbea7e8 in mca_rcache_vma_delete () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_rcache_vma.so
#3  0x00007fefbd7e28d6 in mca_mpool_grdma_register () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_mpool_grdma.so
#4  0x00007fefbcd87f6a in mca_btl_openib_register_mem () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_btl_openib.so
#5  0x00007fefbb494039 in _ompi_osc_rdma_register () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#6  0x00007fefbb494623 in ompi_osc_rdma_frag_alloc () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#7  0x00007fefbb49479c in ompi_osc_rdma_lock_try_acquire_exclusive () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#8  0x00007fefbb494954 in ompi_osc_rdma_lock_acquire_exclusive () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#9  0x00007fefbb494ef0 in ompi_osc_rdma_gacc_local () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#10 0x00007fefbb49716b in ompi_osc_rdma_rget_accumulate_internal () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#11 0x00007fefbb497b3c in ompi_osc_rdma_accumulate () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#12 0x00007fefc2e30062 in PMPI_Accumulate () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libmpi.so.0
#13 0x00000000004010ba in main ()
[rvandevaart@drossetti-ivy4 src]$ 
```

```
[rvandevaart@drossetti-ivy5 dbg-nocuda]$ pstack 28334
Thread 3 (Thread 0x7f6420992700 (LWP 28335)):
#0  0x00007f6422e49173 in epoll_wait () from /lib64/libc.so.6
#1  0x00007f64227cc506 in epoll_dispatch () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#2  0x00007f64227d1cc6 in opal_libevent2022_event_base_loop () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#3  0x00007f64209b444d in progress_engine () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libpmix.so.0
#4  0x00007f64230fb9d1 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f6422e48b7d in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7f641f991700 (LWP 28336)):
#0  0x00007f6422e3f353 in poll () from /lib64/libc.so.6
#1  0x00007f64227d7d1a in poll_dispatch () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#2  0x00007f64227d1cc6 in opal_libevent2022_event_base_loop () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#3  0x00007f6422776b40 in progress_engine () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libopen-pal.so.0
#4  0x00007f64230fb9d1 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f6422e48b7d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f642389f700 (LWP 28334)):
#0  0x00007f641e181cc2 in opal_list_remove_item () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_rcache_vma.so
#1  0x00007f641e183841 in mca_rcache_vma_tree_delete () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_rcache_vma.so
#2  0x00007f641e1817e8 in mca_rcache_vma_delete () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_rcache_vma.so
#3  0x00007f641dd798d6 in mca_mpool_grdma_register () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_mpool_grdma.so
#4  0x00007f641d31ef6a in mca_btl_openib_register_mem () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_btl_openib.so
#5  0x00007f641ba2d039 in _ompi_osc_rdma_register () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#6  0x00007f641ba2d623 in ompi_osc_rdma_frag_alloc () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#7  0x00007f641ba2d79c in ompi_osc_rdma_lock_try_acquire_exclusive () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#8  0x00007f641ba2d954 in ompi_osc_rdma_lock_acquire_exclusive () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#9  0x00007f641ba2e70c in ompi_osc_rdma_gacc_contig () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#10 0x00007f641ba2ed71 in ompi_osc_rdma_gacc_master () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#11 0x00007f641ba301d9 in ompi_osc_rdma_rget_accumulate_internal () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#12 0x00007f641ba30b3c in ompi_osc_rdma_accumulate () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/openmpi/mca_osc_rdma.so
#13 0x00007f64233c6062 in PMPI_Accumulate () from /ivylogin/home/rvandevaart/ompi-repos/ompi-master-rolfv/dbg-nocuda/lib/libmpi.so.0
#14 0x00000000004010ba in main ()
[rvandevaart@drossetti-ivy5 dbg-nocuda]$ 
```

**Comment by hjelmn** [comment]
Haven't seen this. All those tests were running clean with IB. Let me try again and see if something broke.

**Comment by hppritcha** [comment]
Opened #1081 to track the MPI_THREAD_MULTIPLE issue and the flood of libevent messages.

**Comment by hjelmn** [comment]
Can not reproduce the with the latest nightly. How are you building?

**Comment by hjelmn** [comment]
Trying again with more processes since that is what is hanging in your MTT. I still can't get a hang in osc. See an occasional hang in pmix though.

Given the stack trace I think something might be stomping on memory. Will see if I can find any memory errors.

**Comment by rolfv** [comment]
I just walked through this error in the debugger and the error happens when trying to call ibv_post_send() at line 157 in btl_openib_get.c.  We return an error and then keep trying over and over but keep getting the error so it looks like a hang.  Still digging but figured I would report this now.

**Comment by hjelmn** [comment]
Hmm, interesting. What is the error code?

**Comment by rolfv** [comment]
The error code is 22. I am also going to try on another cluster to see if maybe my cluster is the problem.

**Comment by hjelmn** [comment]
Huh, invalid argument. Weird. Let me try running again on a mlx cluster here and see if I get anything similar.

**Comment by hjelmn** [comment]
Opps, did not mean to close

**Comment by rolfv** [comment]
I just reproduced this identical issue on a second cluster. I am getting the error 22 back.

**Comment by rolfv** [comment]
Another data point is that the error happens when the call to mca_btl_openib_get_internal() is made from mca_btl_openib_atomic_internal(). If the call to mca_btl_openib_get_internal() is made from somethere else, we do not get the error.  Specifically, it is from here.
```
[drossetti-ivy4.nvidia.com:08813] Calling mca_btl_openib_get_internal from ../../../../../opal/mca/btl/openib/btl_openib_atomic.c 99
[drossetti-ivy4.nvidia.com:08813] ibv_post_send returned 22
````

**Comment by hjelmn** [comment]
Weird. What does ibv_devinfo -v say?

**Comment by rolfv** [comment]
```
[rvandevaart@drossetti-ivy4 onesided]$ ibv_devinfo -v
hca_id:	mlx5_0
	transport:			InfiniBand (0)
	fw_ver:				10.10.4020
	node_guid:			f452:1403:0026:8390
	sys_image_guid:			f452:1403:0026:8390
	vendor_id:			0x02c9
	vendor_part_id:			4113
	hw_ver:				0x0
	board_id:			MT_1210110019
	phys_port_cnt:			2
	max_mr_size:			0xffffffffffffffff
	page_size_cap:			0xfffffffffffff000
	max_qp:				131072
	max_qp_wr:			16384
	device_cap_flags:		0x00509c36
					BAD_PKEY_CNTR
					BAD_QKEY_CNTR
					AUTO_PATH_MIG
					CHANGE_PHY_PORT
					PORT_ACTIVE_EVENT
					SYS_IMAGE_GUID
					RC_RNR_NAK_GEN
					XRC
					Unknown flags: 0x00408000
	device_cap_exp_flags:		0x5020007100000000
					EXP_DC_TRANSPORT
					EXP_MEM_MGT_EXTENSIONS
					EXP_CROSS_CHANNEL
					EXP_MR_ALLOCATE
					EXT_ATOMICS
					EXT_SEND NOP
					EXP_UMR
	max_sge:			30
	max_sge_rd:			0
	max_cq:				16777216
	max_cqe:			4194303
	max_mr:				16777216
	max_pd:				16777216
	max_qp_rd_atom:			16
	max_ee_rd_atom:			0
	max_res_rd_atom:		2097152
	max_qp_init_rd_atom:		16
	max_ee_init_rd_atom:		0
	atomic_cap:			ATOMIC_HCA_REPLY_BE (64)
	log atomic arg sizes (mask)		3c
	max fetch and add bit boundary	64
	log max atomic inline		5
	max_ee:				0
	max_rdd:			0
	max_mw:				0
	max_raw_ipv6_qp:		0
	max_raw_ethy_qp:		0
	max_mcast_grp:			2097152
	max_mcast_qp_attach:		48
	max_total_mcast_qp_attach:	100663296
	max_ah:				2147483647
	max_fmr:			0
	max_srq:			8388608
	max_srq_wr:			16383
	max_srq_sge:			31
	max_pkeys:			128
	local_ca_ack_delay:		16
	hca_core_clock:			0
	max_klm_list_size:		65536
	max_send_wqe_inline_klms:	20
	max_umr_recursion_depth:	4
	max_umr_stride_dimension:	1
	general_odp_caps:
	rc_odp_caps:
					NO SUPPORT
	uc_odp_caps:
					NO SUPPORT
	ud_odp_caps:
					NO SUPPORT
	dc_odp_caps:
					NO SUPPORT
	xrc_odp_caps:
					NO SUPPORT
	raw_eth_odp_caps:
					NO SUPPORT
	max_dct:			131072
		port:	1
			state:			PORT_ACTIVE (4)
			max_mtu:		4096 (5)
			active_mtu:		4096 (5)
			sm_lid:			8
			port_lid:		1
			port_lmc:		0x00
			link_layer:		InfiniBand
			max_msg_sz:		0x40000000
			port_cap_flags:		0x0651484a
			max_vl_num:		4 (3)
			bad_pkey_cntr:		0x0
			qkey_viol_cntr:		0x0
			sm_sl:			0
			pkey_tbl_len:		128
			gid_tbl_len:		8
			subnet_timeout:		18
			init_type_reply:	0
			active_width:		4X (2)
			active_speed:		14.0 Gbps (16)
			phys_state:		LINK_UP (5)
			GID[  0]:		fe80:0000:0000:0000:f452:1403:0026:8390

		port:	2
			state:			PORT_ACTIVE (4)
			max_mtu:		4096 (5)
			active_mtu:		4096 (5)
			sm_lid:			8
			port_lid:		3
			port_lmc:		0x00
			link_layer:		InfiniBand
			max_msg_sz:		0x40000000
			port_cap_flags:		0x06514848
			max_vl_num:		4 (3)
			bad_pkey_cntr:		0x0
			qkey_viol_cntr:		0x0
			sm_sl:			0
			pkey_tbl_len:		128
			gid_tbl_len:		8
			subnet_timeout:		18
			init_type_reply:	0
			active_width:		4X (2)
			active_speed:		14.0 Gbps (16)
			phys_state:		LINK_UP (5)
			GID[  0]:		fe80:0000:0000:0000:f452:1403:0026:8398

[rvandevaart@drossetti-ivy4 onesided]$ 
```

**Comment by rolfv** [comment]
So this made be obvious, but if I go into opal/include/opal_config.h after running configure but before building and change HAVE_DECL_IBV_ATOMIC_HCA from a 1 to a 0 and then build, the tests pass.

**Comment by hjelmn** [comment]
Yeah, that will disable osc/rdma because it requires the btl to provide atomics.

**Comment by hjelmn** [comment]
Is there any way I can get access to the system? I can not get c_accumulate, etc to fail on my mlx4 systems.

**Comment by hjelmn** [comment]
Ok, this is screwy. I think the mlx5 driver lies about atomic support. From what I can tell the HCA is advertising IBV_ATOMIC_HCA but the driver and libmlx5 both have code what states: "Atomic operations are not supported yet". The problem is that path returns ENOSYS not EINVAL. Will keep digging.

**Comment by hjelmn** [comment]
I talked to Mellanox today. The problem is that with mlx5 the atomic result is returned in network byte order. Because of this they disabled the atomics when using ibv_post_send. It is possible to use the atomics if verbs extensions are installed and ibv_exp_post_send is used. I am coding up a fix to enable the atomics when they are available. Should have it ready in the next day or so.

**Comment by rolfv** [comment]
Interesting and tricky bug to figure out. Thanks for tracking this down Nathan.

**Comment by hjelmn** [comment]
Got it working. osc/rdma should be good to go once #1447 is in 2.0.0.
